README.md
# angular-flapper.js

### Angular-flapper replicates the split-flap displays

This is a directive wrapper of the original jQuery plugin which can be found here: https://github.com/jayKayEss/Flapper.

### Install and Inclusion

As jQuery plugin wrapper, flapper need to be install

Requirements : `bower install flapper`

Add following code in your index.html

```html
<!-- import flapper css style  -->
<link href="bower_components/flapper/css/flapper.css" type="text/css" rel="stylesheet" />

<!-- import flapper jquery  -->
<script src="bower_components/flapper/src/jquery.flapper.js"></script>
```

Next grab angular-flapper with bower : `bower install angular-flapper`

Add following code in your index.html

```html
<!-- import flapper jquery  -->
<script src="bower_components/angular-flapper/dist/ng-flapper.js"></script>
```

Include it in your AngularJS application

```var myApp = angular.module('myApp', ['angular-flapper']);```

### Implementation

```html
<!-- basic implementation -->
<div flapper ng-model="value"  />

<!-- animated implementation -->
<div flapper-repeat ng-model="value"  />

<!-- setting style -->
<div flapper class="slow dark M" ng-model="value"  />

<!-- setting a flapper number -->
<div flapper class="slow dark M" ng-model="value" nb-flap=12 />

<!-- setting a flapper type -->
<div flapper class="slow dark M" ng-model="value" type='num' />

<!-- setting a flapper align -->
<div flapper class="slow dark M" ng-model="value" align='left' />

<!-- maximum timing for digit animation -->
<div flapper class="slow dark M" ng-model="value" timing=250 />

<!-- minimum timing for digit animation -->
<div flapper class="slow dark M" ng-model="value" mintiming=250 />

```

### Theming

For more information, please, see https://github.com/jayKayEss/Flapper.

### Release notes

See CHANGELOG.md for a summary of changes.

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project loosely follows [Semantic Versioning](http://semver.org/),
but is still in early stages of development.


## [0.1.0] - 2016-1129

* Fix README
* Add to bower registery

## [0.0.2] - 2016-1127

* README
* Add options
* Create uglify files
* Github repo is now named `angular-flapper`

## [0.0.1] - 2016-1121

* The module is now named `ngFlapper`


test/karma.conf.js
// Karma configuration
// http://karma-runner.github.io/0.12/config/configuration-file.html
// Generated on 2016-01-28 using
// generator-karma 1.0.1

module.exports = function(config) {
    'use strict';

    config.set({
        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: true,

        // base path, that will be used to resolve files and exclude
        basePath: '../',

        hostname: 'localhost',
        // testing framework to use (jasmine/mocha/qunit/...)
        // as well as any additional frameworks (requirejs/chai/sinon/...)
        frameworks: [
            'jasmine'
        ],


        // list of files / patterns to load in the browser
        files: [
            // bower:js
            'bower_components/jquery/dist/jquery.js',
            'bower_components/angular/angular.js',
            'bower_components/flapper/src/jquery.flapper.js',
            'bower_components/angular-mocks/angular-mocks.js',
            "dist/ng-flapper.js",
            "test/ng-flapper.test.js"
        ],

        // list of files / patterns to exclude
        exclude: [],

        // web server port
        port: 8080,

        // Start these browsers, currently available:
        // - Chrome
        // - ChromeCanary
        // - Firefox
        // - Opera
        // - Safari (only Mac)
        // - PhantomJS
        // - IE (only Windows)
        browsers: [
            "PhantomJS"
        ],

        // Which plugins to enable
        plugins: [
            "karma-phantomjs-launcher",
            "karma-jasmine"
        ],

        // Continuous Integration mode
        // if true, it capture browsers, run tests and exit
        singleRun: false,

        colors: true,

        // level of logging
        // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
        logLevel: config.LOG_WARN,

        // Uncomment the following lines if you are using grunt's server to run the tests
        // proxies: {
        //   '/': 'http://localhost:9000/'
        // },
        // URL root prevent conflicts with the site root
        // urlRoot: '_karma_'
    });
};


test/ng-flapper.test.js
'use strict';

describe('Directive : flapper', function() {

    // load the controller's module
    beforeEach(module('angular-flapper'));

    var $compile, form,
        $rootScope;

    // Store references to $rootScope and $compile
    // so they are available to all tests in this describe block
    beforeEach(inject(function(_$compile_, _$rootScope_) {
        // The injector unwraps the underscores (_) from around the parameter names when matching
        $compile = _$compile_;
        $rootScope = _$rootScope_;
    }));


    it('Replaces the element with the appropriate content', function() {

        // Compile a piece of HTML containing the directive
        form = $compile("<div flapper class='slow dark XS' ng-model='value' nbflap=5 />")($rootScope);

        // fire all the watches, so the scope expression {{1 + 1}} will be evaluated
        $rootScope.$digest();

        // find number od flapper generated
        inject(function($injector) {
            // check the value right after directive attribute
            expect(form.attr('nbflap')).not.toBeUndefined();
            expect(form.attr('nbflap')).toEqual('5');
        });
    });

});


Gruntfile.js
"use strict";
module.exports = function(grunt) {

    // Time how long tasks take. Can help when optimizing build times
    require('time-grunt')(grunt);

    // Automatically load required Grunt tasks
    require('jit-grunt')(grunt, {
        ngtemplates: 'grunt-angular-templates',
        htmllint: 'grunt-html'
    });


    // Configurable paths for the application
    var appConfig = {
        app: 'source',
        dist: 'dist'
    };


    grunt.initConfig({


        // Project settings
        myApp: appConfig,

        pkg: grunt.file.readJSON('package.json'),


        karma: {
            unit: {
                configFile: 'test/karma.conf.js'
            },
            continuous: {
                configFile: 'test/karma.conf.js',
                singleRun: true,
                browsers: ['PhantomJS']
            },
        },

        // Empties folders to start fresh
        clean: {
            dist: {
                files: [{
                    dot: true,
                    src: [
                        '<%= myApp.dist %>/{,*/}*',
                        '!<%= myApp.dist %>/.git{,*/}*',
                        '!<%= myApp.dist %>/bower_components{,*/}*'

                    ]
                }]
            }
        },

        // Analyse JS style 
        jshint: {
            options: {
                jshintrc: '.jshintrc',
                reporter: require('jshint-stylish')
            },
            all: ['Gruntfile.js', '<%= myApp.app %>/scripts/**/*.js', 'server.js', 'server/**/*.js'],

            test: ['<%= myApp.app %>/{,*/}*.test.js']
        },

        // Make sure code styles are up to par
        jscs: {
            options: {
                config: '.jscsrc'
            },
            all: {
                src: [
                    '<%= myApp.app %>/js/{,*/}*.js'
                ]
            },
            test: {
                src: ['<%= myApp.app %>/{,*/}*.test.js']
            }
        },


        jsbeautifier: {
            default: {
                src: ["{,*/}*.js"],
                files: ["dist/ng-flapper.js", "bower_components/{,*/}*.js"]
            }
        },
        // Copies remaining files to places other tasks can use
        copy: {
            dist: {
                files: [{
                    expand: true,
                    dot: true,
                    cwd: '<%= myApp.app %>',
                    dest: '<%= myApp.dist %>',
                    src: [
                        '*.{ico,png,txt,json}',
                        '*.html',
                        'images/{,*/}*.{webp}',
                        'styles/fonts/{,*/}*.*'
                    ]
                }, {
                    expand: true,
                    cwd: '.tmp/images',
                    dest: '<%= myApp.dist %>/images',
                    src: ['generated/*']
                }, {
                    expand: true,
                    cwd: 'bower_components/bootstrap/dist',
                    src: 'fonts/*',
                    dest: '<%= myApp.dist %>'
                }]
            }
        },

        ngAnnotate: {
            options: {
                singleQuotes: true,
            },
            dist: {
                files: {
                    '.tmp/js/app.js': ['<%= myApp.app %>{,*/}*.js']
                }
            }
        },
        // uglify JS and html JS to tmp to dist
        uglify: {
            dist: {
                files: {
                    '<%= myApp.dist %>/ng-flapper.js': [
                        '.tmp/js/app.js'
                    ]
                }
            }
        },
        // Run some tasks in parallel to speed up the build process
        concurrent: {
            dist: [
                'jsbeautifier:default'
            ]
        },
    });

    grunt.registerTask('beautify', ['jsbeautifier:default']);
    grunt.registerTask('test', ['karma:continuous']);

    // tasks apply on all files
    grunt.registerTask('build', [
        'clean:dist',
        'concurrent:dist',
        'copy:dist',
        'ngAnnotate',
        'uglify',
    ]);
};


source/ng-flapper.js
/**
 * ng-Flapper.js v0.0.1
 *
 *
 * Copyright 2016, Clovis OLIVIER
 * Released under the MIT license
 * http://opensource.org/licenses/mit-license.php
 *
 * Date: 28/11/2016
 */

(function(window, document, angular, undefined) {

    "use strict";

    angular.module('angular-flapper', [])


    .directive('flapperRepeat', ['$interval', function($interval) {

        function link(scope, element, attrs) {

            var timeoutId, text;

            element.flapper({
                width: attrs.nbflap,
                format: attrs.format,
                align: attrs.align,
                padding: attrs.padding,
                chars_preset: attrs.type,
                timing: attrs.timing, // the maximum timing for digit animation
                min_timing: attrs.mintiming, // the minimum timing for digit animation
                threshhold: attrs.threshhold, // the point at which Flapper will switch from
                // simple to detailed animations
            });

            scope.$watch('ngModel', function(value) {
                if (value) {
                    text = value;
                }
            });

            element.on('$destroy', function() {
                $interval.cancel(timeoutId);
            });

            setTimeout(function() {
                element.val(text).change();
                var toggle = false;
                timeoutId = setInterval(function() {
                    if (toggle) {
                        element.val(text).change();
                    } else {
                        element.val('').change();
                    }
                    toggle = !toggle;
                }, 5000);
            }, 1000);

        }

        return {
            restrict: 'A',
            transclude: true,
            scope: {
                ngModel: '=',
            },
            link: link
        };
    }])


    .directive('flapper', [function() {

        function link(scope, element, attrs) {

            element.flapper({
                width: attrs.nbflap,
                format: attrs.format,
                align: attrs.align,
                padding: attrs.padding,
                chars_preset: 'alphanum' || attrs.type,
                timing: attrs.timing, // the maximum timing for digit animation
                min_timing: attrs.mintiming, // the minimum timing for digit animation
                threshhold: attrs.threshhold, // the point at which Flapper will switch from
                // simple to detailed animations
            });

            scope.$watch('ngModel', function(value) {
                if (value) {
                    element.val(value).change();
                }
            });

        }

        return {
            restrict: 'A',
            transclude: true,
            scope: {
                ngModel: '=',
            },
            link: link
        };

    }])

})(window, document, angular);


karma.conf.js
// Karma configuration
// http://karma-runner.github.io/0.12/config/configuration-file.html
// Generated on 2016-01-28 using
// generator-karma 1.0.1

module.exports = function(config) {
    'use strict';

    config.set({
        // enable / disable watching file and executing tests whenever any file changes
        autoWatch: true,

        // base path, that will be used to resolve files and exclude
        basePath: './',

        hostname: 'localhost',
        // testing framework to use (jasmine/mocha/qunit/...)
        // as well as any additional frameworks (requirejs/chai/sinon/...)
        frameworks: [
            'jasmine'
        ],


        // list of files / patterns to load in the browser
        files: [
            // bower:js
            'lib/angular/angular.js',
            'lib/flapper/src/jquery.flapper.js',
            'dist/bower_components/angular-mocks/angular-mocks.js',
            // endbower
            'test/{,*/}*.test.js',
        ],

        // list of files / patterns to exclude
        exclude: [],

        // web server port
        port: 8080,

        // Start these browsers, currently available:
        // - Chrome
        // - ChromeCanary
        // - Firefox
        // - Opera
        // - Safari (only Mac)
        // - PhantomJS
        // - IE (only Windows)
        browsers: [
            "PhantomJS"
        ],

        // Which plugins to enable
        plugins: [
            "karma-phantomjs-launcher",
            "karma-jasmine"
        ],

        // Continuous Integration mode
        // if true, it capture browsers, run tests and exit
        singleRun: false,

        colors: true,

        // level of logging
        // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
        logLevel: config.LOG_WARN,

        // Uncomment the following lines if you are using grunt's server to run the tests
        // proxies: {
        //   '/': 'http://localhost:9000/'
        // },
        // URL root prevent conflicts with the site root
        // urlRoot: '_karma_'
    });
};
